src/backend/distributed/planner/multi_explain.c

```diff
@@ -99,10 +173,655 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
 		appendStringInfo(es->str, "%s\n", formattedPlanString);
 	}
 
-	/* if explain printing isn't enabled, print error only after planning */
-	if (!ExplainMultiLogicalPlan && !ExplainMultiPhysicalPlan)
+	if (!ExplainDistributedQueries)
 	{
 		appendStringInfo(es->str, "explain statements for distributed queries ");
+		appendStringInfo(es->str, "are not enabled\n");
+		return;
+	}
+
+	workerJob = multiPlan->workerJob;
+
+	if (list_length(workerJob->dependedJobList) > 0)
+	{
+		appendStringInfo(es->str, "explain statements for complex queries ");
 		appendStringInfo(es->str, "are currently unsupported\n");
+		return;
+	}
+
+	if (es->format == EXPLAIN_FORMAT_TEXT)
+	{
+		appendStringInfoSpaces(es->str, es->indent * 2);
+		appendStringInfo(es->str, "Distributed Query");
+
+		if (multiPlan->masterTableName != NULL)
+		{
+			appendStringInfo(es->str, " into %s", multiPlan->masterTableName);
+		}
+
+		appendStringInfo(es->str, "\n");
+
+		es->indent += 1;
+	}
+
+	ExplainJob(workerJob, es);
+
+	if (es->format == EXPLAIN_FORMAT_TEXT)
+	{
+		es->indent -= 1;
+	}
+
+	routerExecutablePlan = RouterExecutablePlan(multiPlan, TaskExecutorType);
```

Comment:
What does properly mean? A router executor query has no master query, so it's not displayed.